### PR TITLE
Add E2E scenarios for multi process support

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Adds E2E scenarios that verify bugsnag supports multi process apps. This builds on the work in #1019.

## Changeset

Added `multi_process.feature`, which contains the following scenarios in an activity & service that run in different processes:
  - Handled JVM error
  - Unhandled JVM error
  - Handled NDK error
  - Unhandled NDK error
  - SharedPreferences migration for user/device info

Tested by running the mazerunner scenario locally.

### CODE REVIEW

File: features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp [hunk 1]
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
     __builtin_trap();
 }
 
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessHandledCXXErrorScenario_activate(JNIEnv *env,
+                                                                                           jobject instance) {
+  bugsnag_notify_env(env, (char *)"Achtung!",
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:402]: Can we change this to something more meaningful, so if we get payloads spilling or duplicating into other tests we can tell where they originated?

File: features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp [hunk 2]
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
...
+                     (char *)"Das ist nicht gut", BSG_SEVERITY_ERR);
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessUnhandledCXXErrorScenario_user1(JNIEnv *env,
+                                                                                          jobject instance) {
+  bugsnag_set_user_env(env, (char *)"456", (char *)"joe@example.com", (char *)"Joe");
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:409]: As above
```suggestion
  bugsnag_set_user_env(env, (char *)"1", (char *)"1@test.com", (char *)"MultiProcessUnhandledCXXErrorScenario");
```

File: features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp [hunk 3]
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
...
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessUnhandledCXXErrorScenario_user2(JNIEnv *env,
+                                                                                          jobject instance) {
+  bugsnag_set_user_env(env, (char *)"123", (char *)"jane@example.com", (char *)"Jane");
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:415]: As above
```suggestion
  bugsnag_set_user_env(env, (char *)"2", (char *)"2@example.com", (char *)"MultiProcessUnhandledCXXErrorScenario");
```

File: features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/MultiProcessHandledExceptionScenario.kt [hunk 4]
```diff
@@ -20,17 +20,21 @@ internal class MultiProcessHandledExceptionScenario(
...
-            )
-            launchMultiProcessService(scenarioParams::encode)
+            launchMultiProcessService(
+                BugsnagIntentParams.fromConfig(
+                    config,
+                    javaClass.simpleName,
+                    eventMetadata
+                )
+            ) {
+                Bugsnag.setUser("123", "jane@example.com", "Jane")
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/MultiProcessHandledExceptionScenario.kt:31]: Making sure this is more explicit on which user belongs to which process:
```suggestion
                Bugsnag.setUser("2", "background@example.com", "MultiProcessHandledExceptionScenario")
```

File: features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/MultiProcessHandledExceptionScenario.kt [hunk 5]
```diff
@@ -20,17 +20,21 @@ internal class MultiProcessHandledExceptionScenario(
...
+                Bugsnag.notify(generateException())
+            }
+        } else {
+            Bugsnag.setUser("456", "joe@example.com", "Joe")
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/MultiProcessHandledExceptionScenario.kt:35]: As above
```suggestion
            Bugsnag.setUser("2", "foreground@example.com", "MultiProcessHandledExceptionScenario")
```

File: features/full_tests/multi_process.feature [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    Then I wait to receive 2 requests
+    And I sort the requests by "events.0.metaData.process.name"
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "java.lang.RuntimeException"
+    And the exception "message" equals "MultiProcessHandledExceptionScenario"
+    And the event "unhandled" is false
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:14]:
```suggestion
    And the payload field "events.0.user.id" equals "2"
```

File: features/full_tests/multi_process.feature [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.name" equals "Jane"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:15]:
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessHandledExceptionScenario"
```

File: features/full_tests/multi_process.feature [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.email" equals "jane@example.com"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:16]:
```suggestion
    And the payload field "events.0.user.email" equals "background@example.com"
```

File: features/full_tests/multi_process.feature [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "java.lang.RuntimeException"
+    And the exception "message" equals "MultiProcessHandledExceptionScenario"
+    And the event "unhandled" is false
+    And the payload field "events.0.metaData.process.name" equals "com.example.bugsnag.android.mazerunner.multiprocess"
+
+    # device ID is shared between processes
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:28]:
```suggestion
    And the payload field "events.0.user.id" equals "1"
```

File: features/full_tests/multi_process.feature [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.name" equals "Joe"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:29]:
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessHandledExceptionScenario"
```

File: features/full_tests/multi_process.feature [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.email" equals "joe@example.com"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:30]:
```suggestion
    And the payload field "events.0.user.email" equals "foreground@example.com"
```

File: features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp (hunk 1 above)
Comment (Cawllec) [reply features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:402]:
```suggestion
  bugsnag_notify_env(env, (char *)"activate",
```

File: features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp [hunk 12]
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
     __builtin_trap();
 }
 
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessHandledCXXErrorScenario_activate(JNIEnv *env,
+                                                                                           jobject instance) {
+  bugsnag_notify_env(env, (char *)"Achtung!",
+                     (char *)"Das ist nicht gut", BSG_SEVERITY_ERR);
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:403]:
```suggestion
                     (char *)"MultiProcessHandledCXXErrorScenario", BSG_SEVERITY_ERR);
```

File: features/full_tests/multi_process.feature [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.metaData.process.name" equals "com.example.bugsnag.android.mazerunner.multiprocess"
+
+Scenario: Handled NDK error
+    When I run "MultiProcessHandledCXXErrorScenario"
+    Then I wait to receive 2 requests
+    And I sort the requests by "events.0.metaData.process.name"
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "Achtung!"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:63]:
```suggestion
    And the exception "errorClass" equals "activate"
```

File: features/full_tests/multi_process.feature [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    And the exception "message" equals "Das ist nicht gut"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:64]:
```suggestion
    And the exception "message" equals "MultiProcessHandledCXXErrorScenario"
```

File: features/full_tests/multi_process.feature [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    And the event "unhandled" is false
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals the stored value "first_device_id"
+
+    Then I discard the oldest request
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "Achtung!"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:74]:
```suggestion
    And the exception "errorClass" equals "activate"
```

File: features/full_tests/multi_process.feature [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    And the exception "message" equals "Das ist nicht gut"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:75]:
```suggestion
    And the exception "message" equals "MultiProcessHandledCXXErrorScenario"
```

File: features/full_tests/multi_process.feature [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+    And I sort the requests by "events.0.metaData.process.name"
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "SIGTRAP"
+    And the exception "message" equals "Trace/breakpoint trap"
+    And the event "unhandled" is true
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:100]:
```suggestion
    And the payload field "events.0.user.id" equals "2"
```

File: features/full_tests/multi_process.feature [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.name" equals "Jane"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:101]:
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessUnhandledCXXErrorScenario"
```

File: features/full_tests/multi_process.feature [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.email" equals "jane@example.com"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:102]:
```suggestion
    And the payload field "events.0.user.email" equals "2@example.com"
```

File: features/full_tests/multi_process.feature [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+    Then I discard the oldest request
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "SIGTRAP"
+    And the exception "message" equals "Trace/breakpoint trap"
+    And the event "unhandled" is true
+    And the payload field "events.0.metaData.process.name" equals "com.example.bugsnag.android.mazerunner.multiprocess"
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:113]:
```suggestion
    And the payload field "events.0.user.id" equals "1"
```

File: features/full_tests/multi_process.feature [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.name" equals "Joe"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:114]:
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessUnhandledCXXErrorScenario"
```

File: features/full_tests/multi_process.feature [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    And the payload field "events.0.user.email" equals "joe@example.com"
```
Comment (Cawllec) [inline features/full_tests/multi_process.feature:115]:
```suggestion
    And the payload field "events.0.user.email" equals "1@example.com"
```

File: features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/SharedPrefMigrationScenario.kt [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+     * Writes user/device information to the legacy SharedPreferences location to test a migration.
+     */
+    @SuppressLint("ApplySharedPref")
+    private fun persistLegacyPrefs() {
+        val prefs = context.getSharedPreferences("com.bugsnag.android", Context.MODE_PRIVATE)
+
+        if (isRunningFromBackgroundService()) {
+            prefs.edit()
+                .putString("install.iud", "267160a7-5cf2-42d4-be21-969f1573ecb0")
+                .putString("user.id", "456")
```
Comment (Cawllec) [inline features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/SharedPrefMigrationScenario.kt:56]: We should change these user details to be more overtly meaningful like the others.

**Review by Cawllec — COMMENTED** [review]
Looks good.  Most of the following comments are about making the information in the users and error messages more meaningful when test failures occur, especially when multiple users are present in the same test.

One question, is the order of the payloads guaranteed when generating them from two processes?

**Comment by fractalwrench** [comment]
>One question, is the order of the payloads guaranteed when generating them from two processes?

PLAT-5559 has been raised to convert each process to deliver to a separate mazerunner endpoint, but this will require some test infrastructure work first. The scenarios have also been written in a way that should mean the request order remains the same.

Addressed all other feedback, rebased, and ran the multiprocess test scenarios locally.

**Review by Cawllec — APPROVED** [review]

**Pull request merged**